v9 docsite: Add partial support browser matrix (#24283)
microsoft/fluentui: apps/public-docsite-v9/src/Concepts/BrowserSupportMatrix.stories.mdx

<Meta title="Concepts/Developer/Browser Support Matrix" />

> 

Reviewer: I think we're missing a preamble to the set the tone for this page. Expressing the challenge of balancing new features with old browsers, delivering good performance to modern browsers, but not breaking or leaving behind existing customers. Communicating that we're striving to find a balance in-between these two opposing forces.

Author: Intro section added

<Meta title="Concepts/Developer/Browser Support Matrix" />

Defining a browser support matrix to account for all different consumer cases is a difficult undertaking.

> Browser matrix support takes into account various factors such as ECMAScript (ES), browser, and CSS

Reviewer: ```suggestion
This browser matrix takes into account various factors such as ECMAScript (ES), browser, and CSS
```

feature compatibility across different browsers and across different versions of those browsers.

> FluentUI strives to find a balance that enables consumers who support modern browsers to reap the

Reviewer: ```suggestion
Fluent UI strives to find a balance that enables consumers who support modern browsers to reap the
```

> full benefits of the library while at the same time providing the necessary guidance to consumers who

Reviewer: ```suggestion
full benefits of these modern browsers while at the same time providing the necessary guidance to consumers who
```

support older browsers. Details of a full and partial browser support matrix as well as actions consumers

> will need to take such as transpiling and polyfilling are outlined below.

Reviewer: ```suggestion
can take such as transpiling and polyfilling are outlined below.
```

> will need to take such as transpiling and polyfilling are outlined below.

Reviewer: In terms of voice here, we should avoid "will need to take", "must take", etc. Instead, it's preferred to say "can take" "can take to mitigate x", etc.

## Full Browser Support Matrix

> FluentUI fully supports browsers that are at or above the indicated versions below. Consumers who support these browsers

Reviewer: ```suggestion
Fluent UI fully supports browsers that are at or above the indicated versions below. Consumers who support these browsers
```

> are poised to experience the full performance and bundlesize benefits that the FluentUI library provides.

Reviewer: ```suggestion
are poised to experience the full performance and bundlesize benefits that these modern browsers afford the Fluent UI library.
```

| >=14.5        | >=84               | >=16    |

FluentUI will ensure that no browser, EcmaScript(ES), or CSS features that are incompatible with the support matrix laid out above are introduced.
FluentUI will not be providing our own polyfills for features that we expect our browser support matrix to already support.
There are often well-tested third-party polyfills available that applications can use to fill in the gaps.
Consumers who require the use/support of older browsers that don't support certain browser, ES, or CSS features (i.e. ResizeObserver),
will need to implement or use a third-party polyfill to provide the functionality. See the partial support section below for features
that may need to be polyfilled.

## Partial Support:

> FluentUI provides partial support for the browsers below which will have limitations that will require consumers to transpile down and/or use polyfills to cover the support gaps.

Reviewer 2: _Continuing https://github.com/microsoft/fluentui/issues/24721#issuecomment-1245605612_

@JustSlone with that **Fluent UI** transpiles down/uses polyfill/writes custom code to avoid cases when an app throws ➡️ customers that don't need to support browsers the table below are paying bundle size penalties.

Is my understanding correct?

Author: I've added an `Expectations` portion to the doc that should clarify things [a944619](https://github.com/microsoft/fluentui/pull/24283/commits/a944619311ae0d4b244f4f8877c37f4fa53792fb)

**Desktop Browsers:**

| Edge | Firefox | Chrome | Safari | Opera | Internet Explorer |
| ---- | ------- | ------ | ------ | ----- | ----------------- |
| >=79 | >=69    | >=79   | >=13.1 | >=64  | Not Supported     |

**Mobile Browsers:**

| Safari on iOS | Chrome for Android | Samsung |
| ------------- | ------------------ | ------- |
| >=13.4        | >=79               | >=14    |

List of features that are currently used in FluentUI that aren't supported out of the box by the partial support matrix above:

**Unsupported ECMAScript Features:**

> 1. Nullish Coalescing: Unsupported by Chrome, Edge, Firefox, Opera

Reviewer 3: **thought:** AFAIK this could not be polyfilled as its a transpile time transforms - meaning partners will need to manually down-transpile our packages within their tool chain. Is that intended ? if yes can we add it for better context to this document ?

Author: Yeah it's intended that they'll need to transpile down our packages themselves if needed, i'll update the document to make this clearer.

> use/support of older browsers that don't support certain browser, ES, or CSS features (i.e. `ResizeObserver`),

Reviewer: I think the paragraph on polyfills should read something more like this (can't do a suggestion due to the lines not being changed).

"FluentUI will not be providing our own polyfills for features that we expect our browser support matrix to already support. There are often well-tested third-party polyfills available that applications can use to fill in the gaps. Consumers who require the use/support of older browsers that don't support certain browser, ES, or CSS features (i.e. `ResizeObserver`), will need to implement or use a third-party polyfill to provide the functionality. See the partial support section below for features that may need to be polyfilled"

| >=14.5        | >=84               | >=16    |

> FluentUI will ensure that no browser, EcmaScript(ES), or CSS features that are incompatible with the support matrix laid out above are introduced.

Reviewer: ```suggestion
FluentUI will ensure that no browser, ECMAScript(ES), or CSS features that are incompatible with the support matrix laid out above are introduced.
```

FluentUI will not be providing our own polyfills for features that we expect our browser support matrix to already support.
There are often well-tested third-party polyfills available that applications can use to fill in the gaps.
Consumers who require the use/support of older browsers that don't support certain browser, ES, or CSS features (i.e. ResizeObserver),
will need to implement or use a third-party polyfill to provide the functionality. See the partial support section below for features

> that may need to be polyfilled.

Reviewer: It seems strange to start talking about Polyfills here in the full support matrix, shouldn't we talk about it after introducing the partial support, where polyfills start to be expected (and of course if you need support beyond partial you would need to hopefully be able to polyfill further)

Reviewer: I'd suggest having a proper section on polyfills down below the partial browser matrix

Author: Added a `Polyfills` section in the document

FluentUI provides partial support for the browsers below which will have limitations that will require consumers to transpile down and/or use polyfills to cover the support gaps.

**Desktop Browsers:**

| Edge | Firefox | Chrome | Safari | Opera | Internet Explorer |
| ---- | ------- | ------ | ------ | ----- | ----------------- |
| >=79 | >=69    | >=79   | >=13.1 | >=64  | Not Supported     |

**Mobile Browsers:**

| Safari on iOS | Chrome for Android | Samsung |
| ------------- | ------------------ | ------- |
| >=13.4        | >=79               | >=14    |

List of features that are currently used in FluentUI that aren't supported out of the box by the partial support matrix above:

**Unsupported CSS Features:**

1.  Flex gap: Unsupported by Chrome, Edge, Firefox, Opera, Safari, Safari on IOS, Chrome on Android
2.  CSS min, max, clamp: Unsupported by Firefox and Opera
3.  CSS revert value: Unsupported by Firefox and Opera

**Unsupported ECMAScript Features:**
As outlined in the very beginning, FluentUI will be using ES2020 as its Typescript target; so ES2020 features such as nullish coalescing and optional chaining are already being utilized by the library.

> These ES2020 features are not guaranteed to work in all browsers within the partial support matrix so consumers will need to perform down-level transpilation of the library.

Reviewer: Is there a relation or intersection between partial browser support and ECMAScript? I feel like our ECMAScript support is orthogonal to the browser support matrix since we expect apps to transpile to whatever specific browser they need to.

If that's accurate, I would pull this out into the ECMAScript section

Author: I condensed this section a bit but decided to keep it to make it clear that in the case where a consumer wants to use the ES2020 bundle we ship and for some reason decides to follow the partial support matrix that they know it won't work.

| ------------- | ------------------ | ------- |
| >=14.5        | >=84               | >=16    |

> FluentUI will ensure that no browser, ECMAScript, or CSS features that are incompatible with the support matrix laid out above are introduced.

Reviewer: ```suggestion
Fluent UI will ensure that no browser, ECMAScript, or CSS features that are incompatible with the support matrix laid out above are introduced.
```

For consumers who require the use/support of older browsers, please see the partial support matrix and polyfill sections below for more information

> on what will need to be done.

Reviewer: [just a comment] This paragraph is a great improvement, see how it allows the reader to finish at this point. That is useful for the reader to solidity what we've just described as an independent concept. This serves kind of a bookmark in their understanding of the topic, now we are going to build on and elaborate on this core concept further.

## Partial Browser Support Matrix:

> FluentUI introduces the concept of a partial browser support matrix as a means to find that balance between supporting modern browsers and older browsers.

Reviewer: ```suggestion
FluentUI introduces a new concept of a partial browser support matrix as a means to find that balance between supporting modern browsers and older browsers.
```

> The partial browser matrix below will have limitations that will require consumers to transpile down and/or use polyfills to cover the support gaps. The feature gaps

Reviewer: ```suggestion
The partial browser matrix below will have limitations that require consumers to transpile down and/or use polyfills to cover the support gaps. The feature gaps
```

Repeated "will"

I know I commented earlier about saying "can", rather than "must", but in this case this seems OK to me. Perhaps because it's more of a requirement.

are listed below with subsequent recommendation on how to handle each case.

**Desktop Browsers:**

| Edge | Firefox | Chrome | Safari | Opera | Internet Explorer |
| ---- | ------- | ------ | ------ | ----- | ----------------- |
| >=79 | >=69    | >=79   | >=13.1 | >=64  | Not Supported     |

**Mobile Browsers:**

| Safari on iOS | Chrome for Android | Samsung |
| ------------- | ------------------ | ------- |
| >=13.4        | >=79               | >=14    |

> List of features that are currently used in FluentUI that aren't supported out of the box by the partial support matrix above:

Reviewer: ```suggestion
List of features that are currently used in Fluent UI that aren't supported out of the box by the partial support matrix above:
```

I believe Fluent UI should always have a space. You can also say Fluent UI React if you want to be specific.

**Unsupported CSS Features:**

> - Support for the CSS features below are lost within the partial support matrix. The lost of these features will not crash consumer applications,

Reviewer: ```suggestion
- CSS features below are not supported by some browsers within the partial support matrix. The absence of these features will not crash consumer applications,
```

>   it will simply degrade the user interface.

Reviewer: We should link to some screenshots that show what this looks like. It's fine to file an issue on yourself to update this later.

Author: Filed separate issue:https://github.com/microsoft/fluentui/issues/25261

1.  Flex gap: Unsupported by Chrome, Edge, Firefox, Opera, Safari, Safari on IOS, Chrome on Android
2.  CSS min, max, clamp: Unsupported by Firefox and Opera
3.  CSS revert value: Unsupported by Firefox and Opera

> 

Reviewer: Should we link to caniuse for the actual browser support for each of these features below?

Author: [f6bcf72](https://github.com/microsoft/fluentui/pull/24283/commits/f6bcf725dfb782001f72f2d4a8085e625a02f2b7)

**Unsupported ECMAScript Features:**

- ES2020 features listed below are already being used within the FluentUI library. Consumers who follow the partial support matrix will need to transpile down

>   the FluentUI library to make it work since the use of these ES2020 features will result in the application crashing on some of the listed browsers.

Reviewer: ```suggestion
  the Fluent UI library to make it work since the use of these ES2020 features will result in the application crashing on some of the listed browsers.
```

1. Nullish Coalescing: Unsupported by Chrome, Edge, Firefox, Opera.
2. Optional Chaining: Unsupported by Chrome, Edge, Firefox, Opera.

> ## ECMAScript Target & Transpilation

Reviewer: Is this section duplicative of the above section? Should we merge them?

Author: I did think that it was a little repetitive but in the case where someone wants to just quickly glance at the doc and is looking for ECMAScript info specifically, I thought having its own dedicated section was necessary.

Author: Pulled this section out and reiterated ES2020 target within full browser support section and partial browser support section: 
06e733b

> FluentUI will be targetting ES2020 as its Typescript target. FluentUI will not be performing any down-level transpilation below ES2020

Reviewer: ```suggestion
Fluent UI will be targetting ES2020 as its Typescript target. FluentUI will not be performing any down-level transpilation below ES2020
```

> which means that consumers who need to support older browsers (such as those that only support up to ES6) are responsible for transpiling down the FluentUI

Reviewer: ```suggestion
which means that consumers who need to support older browsers (such as those that only support up to ES6) are responsible for transpiling down the Fluent UI
```

library to fit their needs. This ensures that consumers who support modern browsers are not penalized and can get the added benefit of a reduced bundle size that
the latest ES standards provide.

## Polyfills

FluentUI will not be providing polyfills for features we expect our full browser support matrix to already support.

> The only instance that a polyfill may be provided is when FluentUI's use of a feature causes an application to crash on

Reviewer: ```suggestion
The only instance that a polyfill may be provided is when Fluent UI's use of a feature causes an application to crash on
```

the partial support browser matrix. Other than that, consumers will need to implement their own polyfills or use a third party one for their
applications. Handling polyfills on the application level ensures that those on modern browsers don't suffer from unnecessary overhead of

> polyfills being shipped with the FluentUI library by default; as well as prevents any unintended polyfill duplication from the library and app level.

Reviewer: ```suggestion
polyfills being shipped with the Fluent UI library by default; as well as prevents any unintended polyfill duplication from the library and app level.
```

This also gives consumers the flexibility to provide robust solutions such as shipping a bundle with the necessary polyfills to support older browsers and one with a lighter,
polyfill-free bundle to support modern browsers.

## Browser Support Expectations:

**Consumer Expectations:**

> 1. Consumers will need to perform down-level transpilation of the FluentUI library should they need to support browsers that fall below the full support matrix.

Reviewer: ```suggestion
1. Consumers will need to perform down-level transpilation of the Fluent UI library should they need to support browsers that fall below the full support matrix.
```

2. Consumers will need to work around browser, ECMAScript and CSS feature gaps caused by needing to support browsers that fall below the full support matrix.

**Developer Expectations:**

> 1. FluentUI developers and contributors are internally expected to develop against the partial browser support matrix.

Reviewer: ```suggestion
1. Fluent UI developers and contributors are internally expected to develop against the partial browser support matrix.
```

   This means that no code which causes applications to crash on the partially supported browsers (with the exception of ES syntax which are transpilable by the consumer)

>    should be present in the library. Should any application breaking errors be discovered in the future, FluentUI will be responsible for fixing it internally.

Reviewer: ```suggestion
   should be present in the library. Should any application breaking errors be discovered in the future, Fluent UI will be responsible for fixing it internally.
```

> 2. FluentUI developers and contributors reserve the right to use features that fall in between the partial and full browser support matrices (like flex gap) as

Reviewer: ```suggestion
2. Fluent UI developers and contributors reserve the right to use features that fall in between the partial and full browser support matrices (like flex gap) as
```

3.  [CSS revert value](https://caniuse.com/css-revert-value): Unsupported by Opera

**Unsupported ECMAScript Features:**

- As previously mentioned, Fluent UI will be targetting ES2020 as its Typescript target. Some ES2020 features are already present
  in the library and the partial support matrix does not fully support ES2020. This means that consumers who follow the partial support matrix or any browser versions below it
  will need to transpile down the Fluent UI library to make it work since the use of these ES2020 features results in the application crashing on the listed browsers below.

1. Nullish Coalescing: Unsupported by Chrome, Edge, Firefox, Opera.
2. Optional Chaining: Unsupported by Chrome, Edge, Firefox, Opera.

## Polyfills

By default Fluent UI will not be providing polyfills for features we expect our full browser support matrix to already support.
The only instance that a polyfill may be provided is when Fluent UI's use of a feature causes an application to crash on
the partial support browser matrix. Other than that, consumers will need to implement their own polyfills or use a third party one for their

> applications. Handling polyfills on the application level ensures that those on modern browsers don't suffer from unnecessary overhead of

Reviewer 3: this is great explanation. might be good idea to structure this to bullet points so its more "explicit"

```
Other than that, consumers will need to implement their own polyfills or use a third party one for their
applications.

**Why?:**

-  Handling polyfills on the application level ensures that those on modern browsers don't suffer from unnecessary o ...
- prevents any unintended polyfill duplication from the library and app level
- gives consumers the flexibility to provide robust solutions such as shipping a bundle with the necessary polyfills to support older browsers and one with a lighter,
polyfill-free bundle to support modern browsers
```

polyfills being shipped with the Fluent UI library by default; as well as prevents any unintended polyfill duplication from the library and app level.
This also gives consumers the flexibility to provide robust solutions such as shipping a bundle with the necessary polyfills to support older browsers and one with a lighter,
polyfill-free bundle to support modern browsers.

## Developer Expectations:

1. Fluent UI developers and contributors are internally expected to develop against the partial browser support matrix.
   This means that no code which causes applications to crash on the partially supported browsers (with the exception of ES syntax which are transpilable by the consumer)
   should be present in the library. Should any application breaking errors be discovered in the future, Fluent UI will be responsible for fixing it internally.

> 2. Since Fluent UI will be transpiling down and shipping an ES2020 bundle, Fluent UI developers will internally develop against the latest ES standards.

Reviewer 3: ```suggestion
2. Since Fluent UI will be transpiling down and shipping an ES2020, Fluent UI developers will internally develop against the latest ECMA standards.
```
